### PR TITLE
Bug 1131745 - Create testdroid proxy

### PR DESCRIPTION
**Pull request by gregarndt**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+RUN apt-get install -y curl
+
+# Add PPA for latest nodejs versions. Do not need to run apt-get update after this
+# as the script already does it.
+RUN curl -sL https://deb.nodesource.com/setup | sudo bash -
+
+RUN apt-get install -y nodejs python build-essential
+
+RUN npm install -g 6to5
```
Comment (lightsofapollo) [inline Dockerfile]: why here and in npm?

File: src/util.js [hunk 2]
```diff
@@ -0,0 +1,6 @@
+export async function sleep(duration) {
```
Comment (lightsofapollo) [inline src/util.js]: ?

File: src/util.js [hunk 3]
```diff
@@ -0,0 +1,6 @@
...
+  return new Promise(function(accept) {
+    setTimeout(accept, duration);
```
Comment (lightsofapollo) [inline src/util.js]: also note one liner is not terrible here:

``` js
await new Promise(accept => setTimeout(accept, time))
```

File: src/handlers/device.js [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+   * session, a request for a session will be attempted 3 times with a 2 second delay
+   * between attempts.
+   *
+   * @param {Array} devices - List of devices provided by the testdroid api.
+   *
+   * @returns {Object} session
+   */
+  async getDeviceSession(devices) {
+    let session;
+    for(let i = 0; i < devices.length; i++) {
```
Comment (lightsofapollo) [inline src/handlers/device.js]: you can use for of right?

File: src/handlers/device.js [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+      let maxRetries = 3;
+      while (maxRetries-- > 0) {
+        try {
+          session = await this.client.startDeviceSession(device.id);
+          return session;
+        }
+        catch (e) {
+          debug(`Could not start device session for ${device.id}. Retries left: ${maxRetries}. ${e}`);
+          // Noticed a delay between flashing and starting a device session
+          await sleep(2000);
```
Comment (lightsofapollo) [inline src/handlers/device.js]: is 2s really enough time here?

File: src/handlers/device.js [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+    await this.client.stopDeviceSession(device.session.id);
+  }
+
+  /**
+   * Attempts to retrieve a device of a particular type with a particular build.
+   * Devices are labeled with the build url after flashing and this will find devices
+   * with such a label.
+   *
+   * If no devices can be found matching the build, a device will be flashed.
+   * This operation usually takes 3-5 minutes on average.
```
Comment (lightsofapollo) [inline src/handlers/device.js:118]: /me sadface

File: src/handlers/device.js [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+   *
+   * Once a device can be found, ADB and marionette sessions will be created.
+   *
+   * @param {String} deviceType - Type of device that testdroid is aware of. Example: 't2m flame'
+   * @param {String} buildUrl - Location of the build packge used for flashing the device
+   *
+   * @returns {Object} device - Device object that has session and proxy information.
+   */
+  async getDevice(deviceType, buildUrl) {
+    let t = this.client;
```
Comment (lightsofapollo) [inline src/handlers/device.js]: nit: t (again)

File: src/bin/server.js [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    method: 'GET',
+    path: '/devices',
+    handler: async (request, reply) => {
+      debug('/devices');
+      let devices = await server.app.deviceHandler.getDevices();
+      reply(devices);
+    }
+  },
+  {
+    method: 'GET',
```
Comment (lightsofapollo) [inline src/bin/server.js]: REST nit this should likely be a POST or PUT?

Comment (lightsofapollo) [inline src/bin/server.js]: (certainly not get)

**Comment by gregarndt** [comment]
Fixed PR issues with 0ffb1b0ba7712d165dc9d895b2d18172989f62bc

**Pull request merged**